dot/sync/chain_processor.go

```diff
@@ -196,32 +201,39 @@ func (s *chainProcessor) handleBody(body *types.Body) {
 	}
 }
 
-// handleHeader handles blocks (header+body) included in BlockResponses
+// handleBlock handles blocks (header+body) included in BlockResponses
 func (s *chainProcessor) handleBlock(block *types.Block) error {
 	parent, err := s.blockState.GetHeader(block.Header.ParentHash)
 	if err != nil {
 		return fmt.Errorf("%w: %s", errFailedToGetParent, err)
 	}
 
+	hash := parent.Hash()
+	rt, err := s.blockState.GetRuntime(&hash)
+	if err != nil {
+		return fmt.Errorf("getting runtime for parent hash: %w", err)
+	}
+
+	stateVersion, err := rt.StateVersion()
+	if err != nil {
+		return fmt.Errorf("getting runtime state version: %w", err)
+	}
+
 	s.storageState.Lock()
 	defer s.storageState.Unlock()
 
-	ts, err := s.storageState.TrieState(&parent.StateRoot)
+	ts, err := s.storageState.TrieState(&parent.StateRoot, stateVersion)
 	if err != nil {
 		return err
 	}
 
-	root := ts.MustRoot()
-	if !bytes.Equal(parent.StateRoot[:], root[:]) {
+	// TODO shall we remove this? Both are coming from the internal state
+	rootV0 := ts.MustRoot(trie.V0)
+	if !rootV0.Equal(parent.StateRoot) {
+		// TODO add extra check with rootV1 when v1 is supported
 		panic("parent state root does not match snapshot state root")
 	}
```

Comment:
Opinions on that? Should we remove it? Otherwise we'll have to hash tries for v0 and v1 which might have a performance impact.